### PR TITLE
Simplify BufferedReader to improve DataLoader performance

### PR DESCRIPTION
**Pull request by chenwhql**
### PR types

Performance optimization
### PR changes

Others
### Describe

This PR simplifies the operation of creating and copying GPU Tensor by the underlying BufferedReader. 

During GPU training, the original BufferedReader needs to do the following things to buffer a Tensor:

1. Copy Tensor from the BlockingQueue to the CPU buffer
2. Alloc GPU Tensor memory (Independent StreamWait)
3. Copy Tensor from CPU to CUDA Pinned memory
4. Copy Tensor from CUDA Pinned memory to GPU (StreamSync)

These things are time-consuming and require two Syncs. Even if we process these tasks in parallel through threads, the time cannot be compressed to a very short time. When the training time is very short, much less than the above data processing time, the data reading of DataLoader will still become a bottleneck.

Test environments:
- P40 single card, CPU num 25
- dygraph MobileNetV1, batch size 256
- DataLoader num_workers=10

Test data of the above data processing (Batch Image Input):
```
1. [BlockingQueue -> CPU Buffer] PyReader: ReadNext: copy time: 0.040572 s
2. [GPU Memory Alloc] BufferedReader: ReadAsync: gpu prepare time: 0.562235 s
3. [CPU -> CUDA Pinned] BufferedReader: ReadAsync: gpu copy cpu -> cuda pinned time: 0.105902 s
4. [CUDA Pinned -> GPU] BufferedReader: ReadAsync: gpu copy cuda pinned -> cuda time: 0.022287 s
```

But in fact, copying Tensor to cuda memory does not necessarily need to be done by DataLoader. The necessary DataTransform will be performed during Op calculation. If it is GPU training, we can only return Tensor on CUDA Pinned memory.

Doing so can save the time of two Stream synchronizations and greatly improve training efficiency.
Why must synchronize two times in the original DataLoader? because the calculation process here and training is asynchronous, not waiting for the data may be wrong, but if it is done by Op, because the data read by the DataLoader is the beginning of the entire topological sequence network , There is no need to Sync specifically, and there will be no errors. This is more efficient.

From the following test data, a single training step, the data reading efficiency is increased by about **400 times**, and the training speed is increased by about 16%

#### Local P40 Compare (Batchsize 256): + 16.8%

- Test Data of origin:

```
average latency of 50 steps, skip 0 step:
	Avg: 0.738 s/step
	FPS: 346.757 samples/s
average latency of 50 steps, skip 5 steps:
	Avg: 0.738 s/step
	Min: 0.631 s/step
	Max: 0.751 s/step
	FPS: 346.878 samples/s
```

```
[Epoch 0, batch 10], avg_loss 4.93304, acc_top1 0.01953, acc_top5 0.07812, batch_cost: 0.62935 s, net_t: 0.01716 s, backward_t: 0.00481 s, reader_t: 0.02031 s
[Epoch 0, batch 20], avg_loss 4.73635, acc_top1 0.00781, acc_top5 0.05469, batch_cost: 0.62887 s, net_t: 0.01724 s, backward_t: 0.00499 s, reader_t: 0.02020 s
[Epoch 0, batch 30], avg_loss 4.65680, acc_top1 0.01953, acc_top5 0.07812, batch_cost: 0.62766 s, net_t: 0.01711 s, backward_t: 0.00487 s, reader_t: 0.01990 s
[Epoch 0, batch 40], avg_loss 4.67236, acc_top1 0.01172, acc_top5 0.06641, batch_cost: 0.62676 s, net_t: 0.01545 s, backward_t: 0.00415 s, reader_t: 0.01972 s
[Epoch 0, batch 50], avg_loss 4.54960, acc_top1 0.01953, acc_top5 0.09766, batch_cost: 0.73081 s, net_t: 0.01584 s, backward_t: 0.00419 s, reader_t: 0.12186 s
[Epoch 0, batch 60], avg_loss 4.50596, acc_top1 0.01562, acc_top5 0.11719, batch_cost: 0.73087 s, net_t: 0.01571 s, backward_t: 0.00399 s, reader_t: 0.12242 s
[Epoch 0, batch 70], avg_loss 4.40637, acc_top1 0.03906, acc_top5 0.13672, batch_cost: 0.72815 s, net_t: 0.01587 s, backward_t: 0.00388 s, reader_t: 0.12172 s
[Epoch 0, batch 80], avg_loss 4.39635, acc_top1 0.05469, acc_top5 0.17188, batch_cost: 0.73179 s, net_t: 0.01555 s, backward_t: 0.00410 s, reader_t: 0.12241 s
[Epoch 0, batch 90], avg_loss 4.49055, acc_top1 0.03906, acc_top5 0.13281, batch_cost: 0.73156 s, net_t: 0.01590 s, backward_t: 0.00437 s, reader_t: 0.12434 s
[Epoch 0, batch 100], avg_loss 4.22573, acc_top1 0.05469, acc_top5 0.17188, batch_cost: 0.73142 s, net_t: 0.01592 s, backward_t: 0.00423 s, reader_t: 0.12429 s
```

- Test Data of this PR:

```
average latency of 50 steps, skip 0 step:
	Avg: 0.632 s/step
	FPS: 405.302 samples/s
average latency of 50 steps, skip 5 steps:
	Avg: 0.632 s/step
	Min: 0.628 s/step
	Max: 0.647 s/step
	FPS: 405.379 samples/s
```

```
[Epoch 0, batch 10], avg_loss 4.88844, acc_top1 0.01953, acc_top5 0.08594, batch_cost: 0.62908 s, net_t: 0.20680 s, backward_t: 0.00415 s, reader_t: 0.00031 s
[Epoch 0, batch 20], avg_loss 4.67345, acc_top1 0.02344, acc_top5 0.07031, batch_cost: 0.64242 s, net_t: 0.20697 s, backward_t: 0.00421 s, reader_t: 0.01385 s
[Epoch 0, batch 30], avg_loss 4.70440, acc_top1 0.00781, acc_top5 0.09375, batch_cost: 0.62896 s, net_t: 0.20702 s, backward_t: 0.00421 s, reader_t: 0.00036 s
[Epoch 0, batch 40], avg_loss 4.63014, acc_top1 0.02734, acc_top5 0.08203, batch_cost: 0.62771 s, net_t: 0.20672 s, backward_t: 0.00412 s, reader_t: 0.00035 s
[Epoch 0, batch 50], avg_loss 4.57004, acc_top1 0.01953, acc_top5 0.10938, batch_cost: 0.62895 s, net_t: 0.20712 s, backward_t: 0.00414 s, reader_t: 0.00032 s
[Epoch 0, batch 60], avg_loss 4.52250, acc_top1 0.03125, acc_top5 0.11719, batch_cost: 0.63154 s, net_t: 0.20688 s, backward_t: 0.00448 s, reader_t: 0.00035 s
[Epoch 0, batch 70], avg_loss 4.45112, acc_top1 0.01172, acc_top5 0.15625, batch_cost: 0.63901 s, net_t: 0.20697 s, backward_t: 0.00413 s, reader_t: 0.01154 s
[Epoch 0, batch 80], avg_loss 4.41960, acc_top1 0.04688, acc_top5 0.16797, batch_cost: 0.62847 s, net_t: 0.20702 s, backward_t: 0.00430 s, reader_t: 0.00034 s
[Epoch 0, batch 90], avg_loss 4.48008, acc_top1 0.03906, acc_top5 0.16016, batch_cost: 0.64134 s, net_t: 0.20712 s, backward_t: 0.00442 s, reader_t: 0.01047 s
[Epoch 0, batch 100], avg_loss 4.22524, acc_top1 0.04688, acc_top5 0.14062, batch_cost: 0.62963 s, net_t: 0.20731 s, backward_t: 0.00520 s, reader_t: 0.00043 s
```

#### V100 Benchmark Compare (Batchsize 128): + 34%

- Test Data of origin:

image

- Test Data of this PR:

```
average latency of 100 steps, skip 0 step:
	Avg: 0.128 s/step
	FPS: 997.512 samples/s
average latency of 100 steps, skip 5 steps:
	Avg: 0.128 s/step
	Min: 0.122 s/step
	Max: 0.136 s/step
	FPS: 998.054 samples/s
```

### CODE REVIEW

File: python/paddle/fluid/tests/unittests/test_multiprocess_reader_exception.py [hunk 1]
```diff
@@ -94,6 +94,7 @@ def __impl__():
                     num = 0
                     try:
                         for data in reader():
+                            print(data)
```
Comment (heavengate) [inline python/paddle/fluid/tests/unittests/test_multiprocess_reader_exception.py]: remove？

File: python/paddle/fluid/tests/unittests/test_multiprocess_reader_exception.py [hunk 2]
```diff
@@ -77,7 +77,7 @@ def __impl__():
                     reader.decorate_sample_generator(
                         decorated_reader,
                         batch_size=batch_size,
-                        places=fluid.cuda_places())
+                        places=fluid.cuda_places(0))
```
Comment (heavengate) [inline python/paddle/fluid/tests/unittests/test_multiprocess_reader_exception.py:80]: only test single device here？

Comment (chenwhql) [reply python/paddle/fluid/tests/unittests/test_multiprocess_reader_exception.py:80]: 这里有点神奇，这个单测原来就有问题，只能在CUDA_VISIBLE_DEVICES配置单个设备时执行成功，因为它exe.run用的是fluid.CUDAPlace(0)，但DataLoader配置的fluid.cuda_places()，这个是根据环境变量返回结果的，如果CUDA_VISIBLE_DEVICES配置了多卡，这里就错了，所以这里修复了一下

至于有没有必要测多卡，我觉得是没有必要的，这个也是为了故意制造异常catch提升覆盖率的单测

File: python/paddle/fluid/tests/unittests/test_multiprocess_reader_exception.py (hunk 1 above)
Comment (chenwhql) [reply python/paddle/fluid/tests/unittests/test_multiprocess_reader_exception.py]: done, thx

File: python/paddle/fluid/tests/unittests/test_multiprocess_reader_exception.py (hunk 2 above)
Comment (heavengate) [reply python/paddle/fluid/tests/unittests/test_multiprocess_reader_exception.py:80]: ok, LGTM

**Review by heavengate — APPROVED** [review]
LGTM

**Review by phlrain — APPROVED** [review]

**Pull request merged**